File: src/particle.cpp
#include "particle_simulation/particle.hpp"


// Constructor
Particle::Particle(double d_x, double d_y, double d_vx, double d_vy, double d_ax, double d_ay, int id)
    : pos(d_x, d_y), vel(d_vx, d_vy), acc(d_ax, d_ay), id_(id) {}

// Update the particle's position based on its velocity and acceleration
void Particle::Update(double d_time_step) {
    // Update velocity based on acceleration
    vel.x() += acc.x() * d_time_step;
    vel.y() += acc.y() * d_time_step;

    // Update position based on velocity
    pos.x() += vel.x() * d_time_step;
    pos.y() += vel.y() * d_time_step;
}

File: react_component/src/components/Particle.ts
export const GRAVITY = 1000;
export const PARTICLE_RADIUS = 2;
export const FRAME_DT = 1/60;

export class Vector2 {
    constructor(public x: number, public y: number) {}

    norm(): number {
        return Math.sqrt(this.x * this.x + this.y * this.y);
    }
}

export class Particle {
    pos: Vector2;
    vel: Vector2;
    acc: Vector2;
    private id_: number;
    createdAt: number;

    constructor(x: number, y: number, vx: number, vy: number, ax: number, ay: number, id: number) {
        this.pos = new Vector2(x, y);
        this.vel = new Vector2(vx, vy);
        this.acc = new Vector2(ax, ay);
        this.id_ = id;
        this.createdAt = performance.now();
    }

    update(timeStep: number): void {
        // Update velocity with gravity and acceleration
        this.vel.x += this.acc.x * timeStep;
        this.vel.y += this.acc.y * timeStep;

        // Update position with velocity
        this.pos.x += this.vel.x * timeStep;
        this.pos.y += this.vel.y * timeStep;
    }

    id(): number {
        return this.id_;
    }

    setId(newId: number): void {
        this.id_ = newId;
    }
} 